File: source/jugglinglab/view/PatternView.java
// PatternView.java
//
// Copyright 2002-2021 Jack Boyce and the Juggling Lab contributors

package jugglinglab.view;

import java.awt.*;
import java.awt.event.*;
import java.io.StringReader;
import java.text.MessageFormat;
import javax.swing.*;
import javax.swing.event.*;

import jugglinglab.core.*;
import jugglinglab.jml.JMLPattern;
import jugglinglab.notation.Pattern;
import jugglinglab.util.*;


// This view provides the ability to edit the textual representation of a pettern.

public class PatternView extends View implements DocumentListener {
    protected AnimationPanel ja;
    protected JSplitPane jsp;
    protected JRadioButton rb_bp;
    protected JLabel bp_edited_icon;
    protected JRadioButton rb_jml;
    protected JTextArea ta;
    protected JButton compile;
    protected JButton revert;
    protected JLabel lab;
    protected boolean text_edited = false;


    public PatternView(Dimension dim) {
        makePanel(dim);
        updateButtons();
    }

    protected void makePanel(Dimension dim) {
        setLayout(new BorderLayout());

        ja = new AnimationPanel();
        ja.setAnimationPanelPreferredSize(dim);

        JPanel controls = new JPanel();
        GridBagLayout gb = new GridBagLayout();
        controls.setLayout(gb);

        JLabel lab_view = new JLabel(guistrings.getString("PatternView_heading"));
        gb.setConstraints(lab_view, JLFunc.constraints(GridBagConstraints.LINE_START, 0, 0,
                          new Insets(15, 0, 10, 0)));
        controls.add(lab_view);

        ButtonGroup bg = new ButtonGroup();
        JPanel bppanel = new JPanel();
        bppanel.setLayout(new FlowLayout(FlowLayout.LEFT, 0, 0));
        rb_bp = new JRadioButton(guistrings.getString("PatternView_rb1_default"));
        bg.add(rb_bp);
        bppanel.add(rb_bp);
        java.net.URL url = PatternView.class.getResource("/alert.png");
        if (url != null) {
            ImageIcon edited_icon = new ImageIcon(url);
            if (edited_icon != null) {
                ImageIcon edited_icon_scaled = new ImageIcon(edited_icon.getImage().getScaledInstance(22, 22,  java.awt.Image.SCALE_SMOOTH));
                bp_edited_icon = new JLabel(edited_icon_scaled);
                bp_edited_icon.setToolTipText(guistrings.getString("PatternView_alert"));
                bppanel.add(Box.createHorizontalStrut(10));
                bppanel.add(bp_edited_icon);
            }
        }
        controls.add(bppanel);
        gb.setConstraints(bppanel, JLFunc.constraints(GridBagConstraints.LINE_START, 0, 1));

        rb_jml = new JRadioButton(guistrings.getString("PatternView_rb2"));
        bg.add(rb_jml);
        controls.add(rb_jml);
        gb.setConstraints(rb_jml, JLFunc.constraints(GridBagConstraints.LINE_START, 0, 2));

        ta = new JTextArea();
        JScrollPane jscroll = new JScrollPane(ta);
        jscroll.setPreferredSize(new Dimension(400, 1));
        jscroll.setVerticalScrollBarPolicy(JScrollPane.VERTICAL_SCROLLBAR_ALWAYS);
        jscroll.setHorizontalScrollBarPolicy(JScrollPane.HORIZONTAL_SCROLLBAR_ALWAYS);
        controls.add(jscroll);
        GridBagConstraints gbc = JLFunc.constraints(GridBagConstraints.LINE_START, 0, 3,
                                                    new Insets(15, 0, 0, 0));
        gbc.fill = GridBagConstraints.BOTH;
        gbc.weightx = gbc.weighty = 1.0;
        gb.setConstraints(jscroll, gbc);

        jsp = new JSplitPane(JSplitPane.HORIZONTAL_SPLIT, true, ja, controls);
        add(jsp, BorderLayout.CENTER);

        JPanel lower = new JPanel();
        lower.setLayout(new FlowLayout(FlowLayout.LEADING));
        compile = new JButton(guistrings.getString("PatternView_compile_button"));
        lower.add(compile);
        revert = new JButton(guistrings.getString("PatternView_revert_button"));
        lower.add(revert);
        lab = new JLabel("");
        lower.add(lab);
        add(lower, BorderLayout.PAGE_END);

        // add actions to the various items

        ta.getDocument().addDocumentListener(this);

        rb_bp.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent ae) { reloadTextArea(); }
        });

        rb_jml.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent ae) { reloadTextArea(); }
        });

        compile.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent ae) { compilePattern(); }
        });

        revert.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent ae) { revertPattern(); }
        });
    }

    // Update the button configs when a radio button is pressed, the base
    // pattern or JML pattern changes, or we start/stop writing an animated GIF.
    protected void updateButtons() {
        if (ja != null && ja.writingGIF) {
            // writing a GIF
            rb_bp.setEnabled(false);
            rb_jml.setEnabled(false);
            compile.setEnabled(false);
            revert.setEnabled(false);
            return;
        }

        JMLPattern pat = getPattern();

        if (pat == null) {
            rb_bp.setEnabled(false);
            if (bp_edited_icon != null)
                bp_edited_icon.setVisible(false);
            rb_jml.setEnabled(false);
        } else if (pat.getBasePatternNotation() == null || pat.getBasePatternConfig() == null) {
            // no base pattern set
            rb_bp.setEnabled(false);
            if (bp_edited_icon != null)
                bp_edited_icon.setVisible(false);
            rb_jml.setEnabled(true);
            rb_jml.setSelected(true);
        } else {
            rb_bp.setEnabled(true);
            if (bp_edited_icon != null)
                bp_edited_icon.setVisible(pat.isBasePatternEdited());
            rb_jml.setEnabled(true);
        }

        if (rb_bp.isSelected()) {
            compile.setEnabled(pat != null && (pat.isBasePatternEdited() || text_edited));
            revert.setEnabled(text_edited);
        } else if (rb_jml.isSelected()) {
            compile.setEnabled(text_edited);
            revert.setEnabled(text_edited);
        }
    }

    // (Re)load the text in the JTextArea from the pattern, overwriting
    // anything that was there.
    protected void reloadTextArea() {
        if (rb_bp.isSelected())
            ta.setText(getPattern().getBasePatternConfig().replace(";", ";\n"));
        else if (rb_jml.isSelected())
            ta.setText(getPattern().toString());

        ta.setCaretPosition(0);
        lab.setText("");
        setTextEdited(false);

        // Note the above always triggers an updateButtons() call, since
        // text_edited is cycled from true (from the setText() calls) to false.
    }

    protected void setTextEdited(boolean edited) {
        if (text_edited != edited) {
            text_edited = edited;
            updateButtons();
        }
    }

    protected void compilePattern() {
        try {
            if (rb_bp.isSelected()) {
                String notation = getPattern().getBasePatternNotation();
                String config = ta.getText().replace("\n", "").trim();
                JMLPattern newpat = JMLPattern.fromBasePattern(notation, config);
                restartView(newpat, null);
                addToUndoList(newpat);
            } else if (rb_jml.isSelected()) {
                JMLPattern newpat = new JMLPattern(new StringReader(ta.getText()));
                restartView(newpat, null);
                addToUndoList(newpat);
            }
        } catch (JuggleExceptionUser jeu) {
            lab.setText(jeu.getMessage());
            setTextEdited(true);
        } catch (JuggleExceptionInternal jei) {
            ErrorDialog.handleFatalException(jei);
            setTextEdited(true);
        }
        ApplicationWindow.updateWindowMenus();
    }

    protected void revertPattern() {
        reloadTextArea();
    }

    // View methods

    @Override
    public void restartView(JMLPattern p, AnimationPrefs c) throws
                    JuggleExceptionUser, JuggleExceptionInternal {
        ja.restartJuggle(p, c);

        if (p != null) {
            String notation = p.getBasePatternNotation();
            String template = guistrings.getString("PatternView_rb1");
            Object[] arg = { notation == null ? "none set" : notation };
            rb_bp.setText(MessageFormat.format(template, arg));

            if (!(rb_bp.isSelected() || rb_jml.isSelected())) {
                if (notation == null)
                    rb_jml.setSelected(true);
                else
                    rb_bp.setSelected(true);
            }

            updateButtons();
            reloadTextArea();
            parent.setTitle(p.getTitle());
        }
    }

    @Override
    public void restartView() throws JuggleExceptionUser, JuggleExceptionInternal {
        ja.restartJuggle();
    }

    @Override
    public Dimension getAnimationPanelSize() {
        return ja.getSize(new Dimension());
    }

    @Override
    public void setAnimationPanelPreferredSize(Dimension d) {
        ja.setAnimationPanelPreferredSize(d);
        jsp.resetToPreferredSizes();
    }

    @Override
    public JMLPattern getPattern() { return ja.getPattern(); }

    @Override
    public AnimationPrefs getAnimationPrefs() { return ja.getAnimationPrefs(); }

    @Override
    public boolean getPaused() { return ja.getPaused(); }

    @Override
    public void setPaused(boolean pause) {
        if (ja.message == null)
            ja.setPaused(pause);
    }

    @Override
    public void disposeView() { ja.disposeAnimation(); }

    @Override
    public void writeGIF() {
        ja.writingGIF = true;
        updateButtons();
        boolean origpause = getPaused();
        setPaused(true);

        Runnable cleanup = new Runnable() {
            @Override
            public void run() {
                setPaused(origpause);
                ja.writingGIF = false;
                updateButtons();
            }
        };

        new View.GIFWriter(ja, cleanup);
    }

    // javax.swing.event.DocumentListener methods

    @Override
    public void insertUpdate(DocumentEvent e) { setTextEdited(true); }

    @Override
    public void removeUpdate(DocumentEvent e) { setTextEdited(true); }

    @Override
    public void changedUpdate(DocumentEvent e) { setTextEdited(true); }
}

File: composeApp/src/jvmMain/kotlin/jugglinglab/view/PatternView.kt
//
// PatternView.kt
//
// This view provides the ability to edit the text representation of a pattern.
//
// Copyright 2002-2026 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.view

import jugglinglab.composeapp.generated.resources.*
import jugglinglab.core.AnimationPrefs
import jugglinglab.core.PatternAnimationState
import jugglinglab.ui.AnimationPanel
import jugglinglab.ui.PatternWindow
import jugglinglab.jml.JmlPattern
import jugglinglab.jml.PatternBuilder
import jugglinglab.util.jlHandleFatalException
import jugglinglab.util.JuggleExceptionInternal
import jugglinglab.util.JuggleExceptionUser
import jugglinglab.util.jlGetImageResource
import jugglinglab.util.jlConstraints
import jugglinglab.util.jlGetStringResource
import androidx.compose.ui.graphics.toAwtImage
import java.awt.*
import java.awt.event.ActionEvent
import javax.swing.*
import javax.swing.event.DocumentEvent
import javax.swing.event.DocumentListener

class PatternView(
    state: PatternAnimationState,
    patternWindow: PatternWindow
) : View(state, patternWindow), DocumentListener {
    private val ja = AnimationPanel(state, onZoom = onZoomChange)
    private lateinit var jsp: JSplitPane
    private lateinit var rbBp: JRadioButton
    private var bpEditedIcon: JLabel? = null
    private lateinit var rbJml: JRadioButton
    private lateinit var ta: JTextArea
    private lateinit var compile: JButton
    private lateinit var revert: JButton
    private lateinit var lab: JLabel
    private var textEdited: Boolean = false

    init {
        makePanel(Dimension(state.prefs.width, state.prefs.height))
        updateButtons()
    }

    private fun makePanel(dim: Dimension) {
        setLayout(BorderLayout())

        // animator on the left
        ja.preferredSize = if (patternWindow.isWindowMaximized) {
            Dimension(patternWindow.width * 3 / 4, 50)
        } else {
            dim
        }
        ja.minimumSize = Dimension(50, 50)

        // controls panel on the right
        val viewLabel = JLabel(jlGetStringResource(Res.string.gui_patternview_heading))

        rbBp = JRadioButton(jlGetStringResource(Res.string.gui_patternview_rb1_default))
        bpEditedIcon = run {
            val composeImage = jlGetImageResource("alert.png")
            val editedIcon = ImageIcon(composeImage.toAwtImage(), "alert.png")
            val editedIconScaled =
                ImageIcon(
                    editedIcon.getImage().getScaledInstance(22, 22, Image.SCALE_SMOOTH)
                )
            JLabel(editedIconScaled).apply {
                setToolTipText(jlGetStringResource(Res.string.gui_patternview_alert))
            }
        }
        val bppanel = JPanel().apply {
            setLayout(FlowLayout(FlowLayout.LEFT, 0, 0))
            add(rbBp)
            add(Box.createHorizontalStrut(10))
            add(bpEditedIcon)
        }

        rbJml = JRadioButton(jlGetStringResource(Res.string.gui_patternview_rb2))
        ButtonGroup().apply {
            add(rbBp)
            add(rbJml)
        }

        ta = JTextArea()
        val jscroll = JScrollPane(ta).apply {
            preferredSize = Dimension(400, 1)
            setVerticalScrollBarPolicy(JScrollPane.VERTICAL_SCROLLBAR_ALWAYS)
            setHorizontalScrollBarPolicy(JScrollPane.HORIZONTAL_SCROLLBAR_ALWAYS)
        }

        val gb = GridBagLayout().apply {
            setConstraints(viewLabel, jlConstraints(GridBagConstraints.LINE_START, 0, 0, Insets(15, 4, 10, 0)))
            setConstraints(bppanel, jlConstraints(GridBagConstraints.LINE_START, 0, 1, Insets(0, 4, 0, 0)))
            setConstraints(rbJml, jlConstraints(GridBagConstraints.LINE_START, 0, 2, Insets(0, 4, 0, 0)))
            setConstraints(jscroll, jlConstraints(GridBagConstraints.LINE_START, 0, 3, Insets(15, 0, 0, 0)).apply {
                fill = GridBagConstraints.BOTH
                weighty = 1.0
                weightx = 1.0
            })
        }
        val controls = JPanel().apply {
            setLayout(gb)
            add(viewLabel)
            add(bppanel)
            add(rbJml)
            add(jscroll)
        }

        // split pane dividing the two
        jsp = JSplitPane(JSplitPane.HORIZONTAL_SPLIT, true, ja, controls)
        jsp.setResizeWeight(0.0) // % extra space allocated to left (animation) side
        add(jsp, BorderLayout.CENTER)

        // button + error message label across the bottom
        compile = JButton(jlGetStringResource(Res.string.gui_patternview_compile_button))
        revert = JButton(jlGetStringResource(Res.string.gui_patternview_revert_button))
        lab = JLabel(" ")
        val gb2 = GridBagLayout().apply {
            setConstraints(compile, jlConstraints(GridBagConstraints.LINE_START, 0, 0, Insets(8, 8, 8, 0)))
            setConstraints(revert, jlConstraints(GridBagConstraints.LINE_START, 1, 0, Insets(8, 5, 8, 12)))
            setConstraints(lab, jlConstraints(GridBagConstraints.LINE_START, 2, 0).apply {
                fill = GridBagConstraints.HORIZONTAL
                weightx = 1.0
            })
        }
        val lower = JPanel().apply {
            setLayout(gb2)
            add(compile)
            add(revert)
            add(lab)
        }

        add(lower, BorderLayout.PAGE_END)

        // add actions to the various items
        ta.document.addDocumentListener(this)
        rbBp.addActionListener { _: ActionEvent? -> reloadTextArea() }
        rbJml.addActionListener { _: ActionEvent? -> reloadTextArea() }
        compile.addActionListener { _: ActionEvent? -> compilePattern() }
        revert.addActionListener { _: ActionEvent? -> revertPattern() }
    }

    // Update the button configs when a radio button is pressed or the base
    // pattern or JML pattern changes.
    private fun updateButtons() {
        if (!state.pattern.hasBasePattern) {
            // no base pattern set
            rbBp.setEnabled(false)
            bpEditedIcon?.isVisible = false
            rbJml.setEnabled(true)
            rbJml.setSelected(true)
        } else {
            rbBp.setEnabled(true)
            bpEditedIcon?.isVisible = state.pattern.isBasePatternEdited
            rbJml.setEnabled(true)
        }

        if (rbBp.isSelected) {
            compile.setEnabled(state.pattern.isBasePatternEdited || textEdited)
            revert.setEnabled(textEdited)
        } else if (rbJml.isSelected) {
            compile.setEnabled(textEdited)
            revert.setEnabled(textEdited)
        }
    }

    // (Re)load the text in the JTextArea from the pattern, overwriting anything
    // that was there.
    private fun reloadTextArea() {
        if (rbBp.isSelected) {
            ta.text = state.pattern.basePatternConfig!!.replace(";", ";\n")
        } else if (rbJml.isSelected) {
            ta.text = state.pattern.toString()
        }

        ta.setCaretPosition(0)
        lab.setText(" ")
        setTextEdited(false)

        // Note the above always triggers an updateButtons() call, since
        // text_edited is cycled from true (from the setText() calls) to false.
    }

    private fun setTextEdited(edited: Boolean) {
        if (textEdited != edited) {
            textEdited = edited
            updateButtons()
        }
    }

    private fun compilePattern() {
        try {
            if (rbBp.isSelected) {
                val notation = state.pattern.basePatternNotation!!
                val config = ta.getText().replace("\n", "").trim { it <= ' ' }
                val newpat = JmlPattern.fromBasePattern(notation, config)
                restartView(newpat, null)
                state.addCurrentToUndoList()
            } else if (rbJml.isSelected) {
                val newpat = JmlPattern.fromJmlString(ta.getText())
                // set the title in the base pattern
                val record = PatternBuilder.fromJmlPattern(newpat)
                record.setTitleString(newpat.title)
                val newpat2 = JmlPattern.fromPatternBuilder(record)
                restartView(newpat2, null)
                state.addCurrentToUndoList()
            }
        } catch (jeu: JuggleExceptionUser) {
            val errorString = jlGetStringResource(Res.string.error)
            lab.setText("$errorString: " + jeu.message)
            setTextEdited(true)
        } catch (jei: JuggleExceptionInternal) {
            jlHandleFatalException(jei)
            setTextEdited(true)
        }
    }

    private fun revertPattern() {
        reloadTextArea()
    }

    //--------------------------------------------------------------------------
    // View methods
    //--------------------------------------------------------------------------

    @Throws(JuggleExceptionUser::class, JuggleExceptionInternal::class)
    override fun restartView(pattern: JmlPattern?, prefs: AnimationPrefs?, coldRestart: Boolean) {
        val sizeChanged = (prefs != null && (prefs.width != state.prefs.width || prefs.height != state.prefs.height))

        ja.restartJuggle(pattern, prefs, coldRestart)
        if (sizeChanged) {
            setAnimationPanelPreferredSize(
                Dimension(state.prefs.width, state.prefs.height)
            )
            jsp.resetToPreferredSizes()
        }
        if (pattern != null) {
            val notation = pattern.basePatternNotation
            val message = if (notation != null) {
                jlGetStringResource(Res.string.gui_patternview_rb1, notation)
            } else {
                jlGetStringResource(Res.string.gui_patternview_rb1_default)
            }
            rbBp.setText(message)

            if (!rbBp.isSelected && !rbJml.isSelected) {
                if (notation == null) {
                    rbJml.setSelected(true)
                } else {
                    rbBp.setSelected(true)
                }
            }

            updateButtons()
            reloadTextArea()
            patternWindow.setTitle(pattern.title)
            patternWindow.updateColorsMenu()
        }
    }

    override val animationPanelSize: Dimension?
        get() = ja.getSize(Dimension())

    override fun setAnimationPanelPreferredSize(d: Dimension) {
        ja.preferredSize = d
    }

    override fun disposeView() {
        ja.disposeAnimation()
    }

    //--------------------------------------------------------------------------
    // javax.swing.event.DocumentListener methods
    //--------------------------------------------------------------------------

    override fun insertUpdate(e: DocumentEvent?) {
        setTextEdited(true)
    }

    override fun removeUpdate(e: DocumentEvent?) {
        setTextEdited(true)
    }

    override fun changedUpdate(e: DocumentEvent?) {
        setTextEdited(true)
    }
}
